### PR TITLE
Check for undefined jsp-file

### LINKED ISSUE #1702: Validate `web.xml` for missing `<jsp-file>` references
Jetty 9 raises errors when a `web.xml` references missing JSPs.

### PR DESCRIPTION
**Pull request by abbiefarr**
Fix #1702 

Unsure about the use of ComponentCore in WebXmlValidator#validateJsp. I think this assumes that all jsp files will be in src/main/webapp, and I don't know if that's a correct assumption.

@elharo @meltsufin

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/JspFileElementTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import static org.junit.Assert.assertEquals;
+
+import org.eclipse.core.resources.IMarker;
+import org.junit.Test;
+
+public class JspFileElementTest {
+
+  @Test
+  public void testJspFileElement() {
+    DocumentLocation location = new DocumentLocation(3,15);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation.test/src/com/google/cloud/tools/eclipse/appengine/validation/JspFileElementTest.java]: nit: space after comma

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/plugin.properties [hunk 2]
```diff
@@ -7,3 +7,4 @@ servlet.marker=Servlet Marker
 maven.plugin.marker=Obsolete Maven App Engine Plugin
 undefined.servlet.marker=Unresolved Servlet Class
 undefined.servlet.mapping=Undefined Servlet Mapping
+undefined.jsp.file=Unresolved JSP File
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/plugin.properties]: undefined.jsp --> unresolved.jsp

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/JspFileElement.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.wst.validation.internal.provisional.core.IMessage;
+
+/**
+ * A <servlet-name> element within parent <servlet-mapping> whose value has not
+ * been defined.
+ */
+public class JspFileElement extends BannedElement {
+
+  private static final String MARKERID =
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/JspFileElement.java]: MARKER_ID

**Comment by elharo** [comment]
See https://stackoverflow.com/questions/1581683/how-do-i-change-in-an-eclipse-web-project-the-webcontent-folder-to-something-dif

The Deployment Assembly property can be used to define one or more WebContent folders, all of which might contain JSP files. Also see `FacetUtil.addWebFacetToBatch1

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 4]
```diff
@@ -164,11 +164,28 @@ private void validateJsp() {
...
+          // Looks in src/main/webapp/WEB-INF
+          if (file.exists()) {
+            continue;
+          }
+          file = root.getFile("WEB-INF/classes/" + jspName).getUnderlyingFile();
+          // Looks in src/main/java
+          if (file.exists()) {
+            continue;
+          }
+          file = root.getFile("WEB-INF/lib/" + jspName).getUnderlyingFile();
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: This isn't quite right: `lib` will contain jars.  It looks like any [JSPs must be in `META-INF/resources`](http://stackoverflow.com/a/11063065/600339), though I'm not sure if this works for Servlet 2.5.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 5]
```diff
@@ -164,11 +164,28 @@ private void validateJsp() {
           Node jspNode = jspList.item(i);
           String jspName = jspNode.getTextContent();
           IFile file = root.getFile(jspName).getUnderlyingFile();
-          if (!file.exists()) {
-            DocumentLocation location = (DocumentLocation) jspNode.getUserData("location");
-            BannedElement element = new JspFileElement(jspName, location, jspName.length());
-            blacklist.add(element);
+          // Looks in src/main/webapp
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: Comment should indicate that these are example locations

**Comment by abbiefarr** [comment]
Ready for another review

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 6]
```diff
@@ -156,30 +155,30 @@ private void validateServletMapping() {
...
+      //    /            -> src/main/webapp
+      // WEB-INF/classes -> src/main/java
+      // WEB-INF/lib     -> src/main/webapp/WEB-INF/lib
       IVirtualFolder root = component.getRootFolder();
       if (root.exists()) {
         NodeList jspList = document.getElementsByTagName("jsp-file");
         for (int i = 0; i < jspList.getLength(); i++) {
           Node jspNode = jspList.item(i);
           String jspName = jspNode.getTextContent();
           IFile file = root.getFile(jspName).getUnderlyingFile();
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: Would separating our the resolve-JSP into a separate method (`IFile resolveJsp(IVirtualComponent, String jspPath)`) simplify this method?

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 7]
```diff
@@ -144,6 +149,47 @@ private void validateServletMapping() {
...
+      //    /            -> src/main/webapp
+      // WEB-INF/classes -> src/main/java
+      // WEB-INF/lib     -> src/main/webapp/WEB-INF/lib
+      IVirtualFolder root = component.getRootFolder();
+      if (root.exists()) {
+        NodeList jspList = document.getElementsByTagName("jsp-file");
+        for (int i = 0; i < jspList.getLength(); i++) {
+          Node jspNode = jspList.item(i);
+          String jspName = jspNode.getTextContent();
+          IFile file = root.getFile(jspName).getUnderlyingFile();
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: Would extracting this jsp-resolving into a new method (`IFile resolveJsp(IVirtualComponent,String)`) might help simplify this method?

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 8]
```diff
@@ -144,6 +149,53 @@ private void validateServletMapping() {
     }
   }
   
+  /**
+   * Verifies that every <jsp-file> element exists in the project.
+   */
+  private void validateJsp() {
+    IProject project = resource.getProject();
+    IVirtualComponent component = ComponentCore.createComponent(project);
+    // TODO: Search for JSP files in jars if web.xml is version 3.0+.
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: todo probably belongs in resolve.jsp and perhaps file an issue on flex for this

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 9]
```diff
@@ -144,6 +149,53 @@ private void validateServletMapping() {
     }
   }
   
+  /**
+   * Verifies that every <jsp-file> element exists in the project.
+   */
+  private void validateJsp() {
+    IProject project = resource.getProject();
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: You should first check if the servlet-version is 2.5. If it isn't, just skip the rest of this.

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 10]
```diff
@@ -144,6 +149,62 @@ private void validateServletMapping() {
...
+    // TODO: Search for JSP files in jars if web.xml is version 3.0+.
+    // JSP file META-INF/resources/test.jsp in a jar should be able to be referenced
+    // as test.jsp in web.xml.
+    return false;
+  }
+  
+  private boolean isVersion30OrLater() {
+    Node node = document.getFirstChild();
+    String versionString = (String) node.getUserData("version");
+    double version = Double.parseDouble(versionString);
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: need to handle possible NumberFormatException here

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 11]
```diff
@@ -144,6 +149,62 @@ private void validateServletMapping() {
...
+    }
+    // TODO: Search for JSP files in jars if web.xml is version 3.0+.
+    // JSP file META-INF/resources/test.jsp in a jar should be able to be referenced
+    // as test.jsp in web.xml.
+    return false;
+  }
+  
+  private boolean isVersion30OrLater() {
+    Node node = document.getFirstChild();
+    String versionString = (String) node.getUserData("version");
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: what if this element isn't present?

File: plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java [hunk 12]
```diff
@@ -144,6 +149,62 @@ private void validateServletMapping() {
...
+    if (file.exists()) {
+      return true;
+    }
+    // TODO: Search for JSP files in jars if web.xml is version 3.0+.
+    // JSP file META-INF/resources/test.jsp in a jar should be able to be referenced
+    // as test.jsp in web.xml.
+    return false;
+  }
+  
+  private boolean isVersion30OrLater() {
```
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.appengine.validation/src/com/google/cloud/tools/eclipse/appengine/validation/WebXmlValidator.java]: Might be safer to check explicitly for version 2.5.

**Pull request merged**